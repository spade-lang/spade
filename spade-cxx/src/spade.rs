use std::ops::Deref;

use color_eyre::eyre::bail;
use color_eyre::eyre::Context;
use color_eyre::eyre::Result;
use color_eyre::owo_colors::OwoColorize;
use cxx::CxxString;
use num::ToPrimitive;

struct CompilerState(pub spade::compiler_state::CompilerState);

impl CompilerState {
    fn list_names(&self) {
        for (from, to) in &self.name_source_map.inner {
            println!("{from} -> {to}")
        }
    }

    fn demangle_name(&self, name: &str) -> String {
        self.0
            .demangle_string(name)
            .unwrap_or(format!("(not demangled) {name}"))
    }
}

fn read_state(path: &str) -> Result<Box<CompilerState>> {
    let file = std::fs::read_to_string(path)
        .with_context(|| "Failed to read compiler state from {path}")?;

    let ron = ron::Options::default().without_recursion_limit();

    Ok(Box::new(CompilerState(
        ron.from_str(&file)
            .context("Failed to decode compiler state in {path}")?,
    )))
}

impl Deref for CompilerState {
    type Target = spade::compiler_state::CompilerState;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

struct SignalValue(pub spade_mir::eval::Value);

impl SignalValue {
    fn as_u64(&self) -> u64 {
        self.0.as_u64()
    }

    fn as_u32_chunks(&self) -> Vec<u32> {
        let mut digits = self.0.as_u32_chunks().to_u32_digits();
        let expected_digit_count = ((self
            .0
            .width()
            .to_u64()
            .expect("Value size does not fit in u64") as f64)
            / 32.)
            .ceil() as u64;

        digits.extend_from_slice(&vec![0; expected_digit_count as usize - digits.len()]);
        digits
    }
}

struct SimulationExt(pub spade_simulation_ext::Spade);
struct BitString(spade_simulation_ext::BitString);

fn new_bit_string(s: &CxxString) -> Box<BitString> {
    Box::new(BitString(spade_simulation_ext::BitString(s.to_string())))
}

struct ComparisonResult(spade_simulation_ext::ComparisonResult);
impl ComparisonResult {
    fn matches(&self) -> bool {
        self.0.matches()
    }
}

struct FieldRef(spade_simulation_ext::field_ref::FieldRef);

fn setup_spade(uut_name: String, state_path: String) -> Result<Box<SimulationExt>> {
    Ok(Box::new(SimulationExt(spade_simulation_ext::Spade::new(
        uut_name, state_path,
    )?)))
}

impl SimulationExt {
    pub fn port_value(&mut self, port: &str, expr: &str) -> Result<Box<SignalValue>> {
        self.0
            .port_value_raw(port, expr)
            .map(|(_, value)| Box::new(SignalValue(value)))
    }

    pub fn compare_field(
        &mut self,
        // The field to compare
        field: &FieldRef,
        // The spade expression to compare against
        spade_expr: &str,
        // The bits of the whole output struct
        output_bits: &BitString,
    ) -> Result<Box<ComparisonResult>> {
        self.0
            .compare_field(field.0.clone(), spade_expr, &output_bits.0)
            .map(|o| Box::new(ComparisonResult(o)))
    }

    pub fn assert_eq(
        &mut self,
        // The field to compare
        field: &FieldRef,
        // The spade expression to compare against
        spade_expr: &str,
        // The bits of the whole output struct
        output_bits: &BitString,
        source_loc: &str,
    ) -> Result<()> {
        let cmp_result = self
            .0
            .compare_field(field.0.clone(), spade_expr, &output_bits.0)?;

        if !cmp_result.matches() {
            println!("{}", source_loc.to_string().red());
            println!("\texpected: {}", cmp_result.expected_spade);
            println!("\tgot:      {}", cmp_result.got_spade);
            println!();
            println!(
                "\tverilog (\n\t'{}' != \n\t'{}')",
                cmp_result.expected_bits.0.green(),
                cmp_result.got_bits.0.red()
            );
            bail!("{source_loc}: assertion failed");
        }
        Ok(())
    }

    pub fn output_field(&mut self, path: &Vec<String>) -> Result<Box<FieldRef>> {
        self.0.output_field(path.clone()).and_then(|o| match o {
            Some(field) => Ok(Box::new(FieldRef(field))),
            None => bail!("Trying to access output field on a unit returning void"),
        })
    }

    pub fn field_value(&mut self, field: &FieldRef, output_bits: &BitString) -> Result<String> {
        self.0.field_value(field.0.clone(), &output_bits.0)
    }
}

#[cxx::bridge(namespace = "spade")]
mod ffi {

    extern "Rust" {
        type FieldRef;
    }

    extern "Rust" {
        type BitString;
    }

    extern "Rust" {
        type ComparisonResult;

        fn matches(&self) -> bool;
    }

    extern "Rust" {
        fn new_bit_string(s: &CxxString) -> Box<BitString>;
    }

    extern "Rust" {
        type SignalValue;

        fn as_u64(&self) -> u64;
        fn as_u32_chunks(&self) -> Vec<u32>;
    }

    extern "Rust" {
        type CompilerState;

        fn read_state(path: &str) -> Result<Box<CompilerState>>;
        fn list_names(&self);
        fn demangle_name(&self, name: &str) -> String;
    }

    extern "Rust" {
        type SimulationExt;

        fn setup_spade(uut_name: String, state_path: String) -> Result<Box<SimulationExt>>;
        fn port_value(&mut self, port: &str, expr: &str) -> Result<Box<SignalValue>>;
        fn compare_field(
            &mut self,
            // The field to compare
            field: &FieldRef,
            // The spade expression to compare against
            spade_expr: &str,
            // The bits of the whole output struct
            output_bits: &BitString,
        ) -> Result<Box<ComparisonResult>>;
        fn assert_eq(
            &mut self,
            // The field to compare
            field: &FieldRef,
            // The spade expression to compare against
            spade_expr: &str,
            // The bits of the whole output struct
            output_bits: &BitString,
            source_loc: &str,
        ) -> Result<()>;

        fn output_field(&mut self, path: &Vec<String>) -> Result<Box<FieldRef>>;

        pub fn field_value(&mut self, field: &FieldRef, output_bits: &BitString) -> Result<String>;
    }
}
