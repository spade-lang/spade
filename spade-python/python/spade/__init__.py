from ast import FunctionDef
from typing import Any, Callable, List, NewType, Tuple
import typing
from .spade import Spade, FieldRef, BitString, SourceCodeError, CompilationError

import cocotb
import colors
from cocotb.types import LogicArray
from spade import Spade

import os
import traceback
from pathlib import Path
import logging
import sys

class HandledAlreadyException(Exception):
    def __init__(self, inner) -> None:
        self.__cause__ = inner


# Cocotb adds a ton of extra chars to lines which makes them unnecessarily long, we'll override
# its logger with one that simply prints the log message
class SimpleFormatter(logging.Formatter):
    def _format(self, msg):
        print(msg)


# FIXME: Once we only support newer python versions, we should use this typedef
# type SpadeConvertible = str | int | bool | List[SpadeConvertible]

def handle_maybe_spade_exception(e: BaseException, stack: list[traceback.FrameSummary]):
    filtered_stack = list(filter(
        lambda frame: frame.filename != __file__ and "site-packages/cocotb" not in frame.filename,
        stack
    ))

    workdir = os.getcwd()
    top = filtered_stack[-1]
    lineno = f"{top.lineno}"
    lineno_padding = "".join(map(lambda _: " ", lineno))
    line_ref = f"{Path(top.filename).relative_to(workdir)}:{top.lineno}"

    def draw_error(kind: str, err_color: Callable[[str], Any], description: str, primary_label: str, notes: list[str]):
        print()

        print(colors.bold(f"{err_color(f'{kind}:')} {description}"))
        print(f" {lineno_padding} {colors.blue('┌─')} {line_ref}")
        print(colors.blue(f" {lineno_padding} │"))
        print(f" {colors.blue(f'{lineno} │')} {top.line}")

        if top.line is not None:
            arrows = err_color(f"{''.join(map(lambda _: '^', top.line))} {primary_label}")
            print(f" {colors.blue(f'{lineno_padding} │ ')}{arrows}", )

        print("")
        for note in notes:
            print(f"  {colors.bold(f'Help: ')}{note}")

        print("")
        if filtered_stack[:-1]:
            print("Traceback:")
            for frame in filtered_stack[:-1]:
                print(f"  {colors.green(Path(frame.filename).relative_to(workdir))}:{frame.lineno}")
        print("")

        # We've printed the exception, we don't need cocotb to do so again. So we'll just turn off 
        # logging
        logging.getLogger().handlers = []
        



    if isinstance(e, HandledAlreadyException):
        raise e
    if isinstance(e, SourceCodeError):
        description = e.args[0].description
        primary_label = e.args[0].primary_label
        notes = e.args[0].notes

        draw_error('error', lambda x: colors.red(x, style='bold'), description, primary_label, notes)
        raise HandledAlreadyException(e)
    if isinstance(e, CompilationError):
        # Spade prints the actual error itself, we just have to tell the user what python
        # line the error originated on
        draw_error("note", colors.green, "A Spade expression failed to compile", "When executing this", [])
        raise HandledAlreadyException(e)

    raise e

    

class SpadeExt(Spade):
    def __new__(cls, dut):
        # logger = logging.getLogger("cocotb").setFormatter()
        hdlr = logging.StreamHandler(sys.stdout)
        hdlr.setFormatter(SimpleFormatter())
        logging.getLogger().handlers = [hdlr]

        compiler_state = os.environ["SWIM_SPADE_STATE"]
        uut_name = os.environ["SWIM_UUT"]

        try:
            result = super().__new__(cls, uut_name, compiler_state)
        except FileNotFoundError as e:
            print(f"{compiler_state}")
            print("Failed to find", e.filename, " ", e.filename2)
            raise e

        result.dut = dut
        result.i = InputPorts(dut, result)
        result.o = Field(result, result.output_as_field_ref(), dut)
        return result


class Field(object):
    def __init__(self, spade: SpadeExt, field_ref: FieldRef, dut):
        self.spade__ = spade
        self.field_ref__ = field_ref
        self.dut__ = dut

    # This is not intended to be called on this struct, instead, it should be called on the
    # parent field since python does not allow overloading operator=
    def set_value__(self, value):
        try:
            value = to_spade_value(value)
            bit_string = self.spade__.compile_field_value(self.field_ref__, value)
            fwd_range = self.field_ref__.write_dir_range()
            bits = LogicArray(bit_string.inner())
            signal = self.dut__._id(self.field_ref__.source.fwd_mangled(), extended=False)
            downto_range = fwd_range.as_downto(len(signal))
            if fwd_range.is_full():
                signal.value = bits
            else:
                for i, idx in enumerate(range(downto_range[0], downto_range[1])):
                    signal[idx].value = bits[i]
        except Exception as e:
                handle_maybe_spade_exception(e, traceback.extract_stack())

    def assert_eq(self, expected: object):
        try:
            expected = to_spade_value(expected)
            # This shares a bit of code with is_eq, but since we need access to intermediate
            # values, we'll duplicate things for now
            r = self.spade__.compare_field(
                self.field_ref__, expected, BitString(self.dut__._id(self.field_ref__.source.back_mangled(), extended=False).value.binstr)
            )

            expected_bits = r.expected_bits.inner()
            got_bits = r.got_bits.inner()

            if not r.matches():
                message = "\n"
                message += colors.red("Assertion failed") + "\n"
                message += f"\t expected: {colors.green(r.expected_spade)}\n"
                message += f"\t      got: {colors.red(r.got_spade)}\n"
                message += "\n"
                message += f"\tverilog ('{colors.green(expected_bits)}' != '{colors.red(got_bits)}')"
                assert False, message
        except Exception as e:
            handle_maybe_spade_exception(e, traceback.extract_stack())

    def value(self):
        """
        Returns the value of the field as a string representation of the spade value.
        """
        # This shares a bit of code with is_eq, but since we need access to intermediate
        # values, we'll duplicate things for now
        return self.spade__.field_value(
            self.field_ref__, BitString(self.dut__.output__.value.binstr)
        )

    def is_eq(self, other: object) -> bool:
        return self == other

    def __ne__(self, value: object, /) -> bool:
        return not self == value

    def __eq__(self, value: object, /) -> bool:
        try:
            value = to_spade_value(typing.cast(object, value))
            r = self.spade__.compare_field(
                self.field_ref__, value, BitString(self.dut__.output__.value.binstr)
            )
            expected_bits = r.expected_bits.inner()
            got_bits = r.got_bits.inner()
            return expected_bits.lower() == got_bits.lower()
        except Exception as e:
                handle_maybe_spade_exception(e, traceback.extract_stack())

    def __getattribute__(self, __name: str):
        if __name.endswith("__") or __name in ["assert_eq", "is_eq", "value"]:
            return super(Field, self).__getattribute__(__name)
        else:
            try:
                return Field(
                    self.spade__,
                    self.spade__.field_of_field(self.field_ref__, __name),
                    self.dut__,
                )
            except Exception as e:
                handle_maybe_spade_exception(e, traceback.extract_stack())


    def __setattr__(self, name: str, value: object):
        if not name.endswith("__"):
            self.__getattribute__(name).set_value__(value)
        else:
            super(Field, self).__setattr__(name, value)


class InputPorts(object):
    def __init__(self, dut, spade: SpadeExt):
        self.spade__ = spade
        self.dut__ = dut

    def __getattribute__(self, __name: str) -> Field:
        if __name.endswith("__"):
            return super(InputPorts, self).__getattribute__(__name)

        try:
            return Field(self.spade__, self.spade__.arg_as_field(__name), self.dut__)
        except Exception as e:
            handle_maybe_spade_exception(e, traceback.extract_stack())

    def __setattr__(self, name: str, value: object):
        if not name.endswith("__"):
            self.__getattribute__(name).set_value__(value)
        else:
            super(InputPorts, self).__setattr__(name, value)


def to_spade_value(val: object) -> str:
    if type(val) == str:
        return val
    elif type(val) == int:
        return f"{val}"
    elif type(val) == bool:
        return "true" if val else "false"
    elif type(val) == list:
        return f"[{', '.join(map(lambda v: to_spade_value(v), val))}]"
    else:
        raise TypeError(
            f"Values of type {type(val)} cannot be converted to Spade values"
        )
