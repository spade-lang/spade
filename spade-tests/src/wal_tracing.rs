use spade::Artefacts;
use spade_common::{location_info::WithLocation, name::Path, num_ext::InfallibleToBigUint};
use spade_hir_lowering::MirLowerable;
use spade_mir::{renaming::VerilogNameSource, types::Type};
use spade_typeinference::{equation::TypedExpression, TypeState};

use crate::{build_artifacts, snapshot_error};

fn get_field_type(artefacts: &Artefacts, target_name: &str) -> Type {
    let (name, _) = artefacts
        .state
        .symtab
        .symtab()
        .lookup_unit(&Path::from_strs(&["main"]).nowhere())
        .unwrap();

    // Look up the mir context of the unit we are observing
    let mir_ctx = artefacts
        .state
        .mir_context
        .get(&name)
        .expect(&format!("Did not find information a unit named `main`"));

    let source = mir_ctx
        .verilog_name_map
        .lookup_name(target_name)
        .expect(&format!(
            "Did not find spade variable for verilog identifier '{target_name}'"
        ));

    let typed_expr = match source {
        VerilogNameSource::ForwardName(n) => TypedExpression::Name(n.clone()),
        VerilogNameSource::ForwardExpr(id) => TypedExpression::Id(*id),
        VerilogNameSource::BackwardName(_) | VerilogNameSource::BackwardExpr(_) => {
            panic!("Translation of backward port types is unsupported")
        }
    };

    let ty = mir_ctx
        .type_map
        .type_of(&typed_expr)
        .expect(&format!("Did not find a type for {typed_expr}"));

    let mir_ty = TypeState::ungenerify_type(
        ty,
        artefacts.state.symtab.symtab(),
        &artefacts.state.item_list.types,
    )
    .expect(&format!("Tried to ungenerify generic type {ty}"))
    .to_mir_type();

    mir_ty
}

// NOTE: This tests do_wal_trace_lowering, in particular some hacky code being performed
// there. If these tests fail, it might be time to clean stuff up
#[test]
fn wal_traced_struct_with_multiple_backward_ports_has_type_information() {
    let code = r#"
        #[wal_traceable(suffix = wal_suffix__)]
        struct port Test {
            a: &int<8>,
            b: inv &int<4>,
            c: &int<16>,
            d: inv &int<7>
        }

        entity main(x: Test) -> Test {
            #[wal_trace]
            let y = x;
            y
        }
    "#;

    let artefacts = build_artifacts(code, true);

    assert_eq!(
        get_field_type(&artefacts, "y__a__wal_suffix__"),
        Type::Int(8u64.to_biguint())
    );
    assert_eq!(
        get_field_type(&artefacts, "y__b__wal_suffix__"),
        Type::Int(4u64.to_biguint())
    );
    assert_eq!(
        get_field_type(&artefacts, "y__c__wal_suffix__"),
        Type::Int(16u64.to_biguint())
    );
    assert_eq!(
        get_field_type(&artefacts, "y__d__wal_suffix__"),
        Type::Int(7u64.to_biguint())
    );
}

#[test]
fn wal_suffixed_variable_has_type_info() {
    let code = r#"
        entity main(x: int<8>) -> int<8> {
            #[wal_suffix(suffix=__wal_suffix__)]
            let y = x;
            y
        }
    "#;

    let artefacts = build_artifacts(code, true);

    assert_eq!(
        get_field_type(&artefacts, "y__wal_suffix__"),
        Type::Int(8u64.to_biguint())
    );
}

#[test]
fn wal_suffixed_struct_does_not_expand() {
    let code = r#"
        #[wal_traceable()]
        struct Test {
            a: int<8>,
        }

        entity main(x: Test) -> Test {
            #[wal_suffix(suffix=__wal_suffix__)]
            let y = x;
            y
        }
    "#;

    let artefacts = build_artifacts(code, true);

    assert_eq!(
        get_field_type(&artefacts, "y__wal_suffix__"),
        Type::Struct(vec![(String::from("a"), Type::Int(8u64.to_biguint()))])
    );
}

#[test]
fn wal_suffix_on_register_works() {
    let code = r#"
        entity main(clk: clock, x: int<8>) -> int<8> {
            #[wal_suffix(suffix=__wal_suffix__)]
            reg(clk) y = x;
            y
        }
    "#;

    let artefacts = build_artifacts(code, true);

    assert_eq!(
        get_field_type(&artefacts, "y__wal_suffix__"),
        Type::Int(8u64.to_biguint())
    );
}

#[test]
fn wal_suffix_on_unit_works() {
    let code = r#"
        #[wal_suffix(suffix=__wal_suffix__)]
        entity main(clk: clock, x: int<8>) -> int<8> {
            x
        }
    "#;

    let artefacts = build_artifacts(code, true);

    assert_eq!(get_field_type(&artefacts, "clk__wal_suffix__"), Type::Bool);
    assert_eq!(
        get_field_type(&artefacts, "x__wal_suffix__"),
        Type::Int(8u64.to_biguint())
    );
}

snapshot_error! {
    wal_suffix_on_extern_unit_is_error,
    r#"
        #[wal_suffix(suffix=__wal_suffix__)]
        extern entity main(clk: clock, x: int<8>) -> int<8>;
    "#
}
