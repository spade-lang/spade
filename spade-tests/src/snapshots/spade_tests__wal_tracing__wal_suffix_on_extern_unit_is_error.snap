---
source: spade-tests/src/wal_tracing.rs
---
#[wal_suffix(suffix=__wal_suffix__)]
extern entity main(clk: clock, x: int<8>) -> int<8>;


error: wal_suffix is not allowed on `extern` units
  ┌─ testinput:1:1
  │
1 │ #[wal_suffix(suffix=__wal_suffix__)]
  │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Not allowed on `extern` units
2 │ extern entity main(clk: clock, x: int<8>) -> int<8>;
  │ ------ This unit is `extern`
